rainier-core/src/main/scala/com/stripe/rainier/compute/Decimal.scala

```diff
@@ -1,47 +1,129 @@
 package com.stripe.rainier.compute
 
-case class Decimal(value: BigDecimal) {
-  def toDouble = value.toDouble
-  def toInt = value.toInt
+import scala.annotation.tailrec
 
-  def isValidInt = value.isValidInt
-  def isWhole = value.isWhole
+sealed trait Decimal {
+  def ==(other: Decimal) =
+    toDouble == other.toDouble
+  override def hashCode(): Int =
+    toDouble.hashCode()
 
-  def abs = Decimal(value.abs)
+  def toDouble: Double
+  def toInt: Int = toDouble.toInt
 
-  def pow(exponent: Int) =
-    Decimal(value.pow(exponent))
+  def isValidInt = toDouble.isValidInt
+  def isWhole = toDouble.isWhole
 
   def >(other: Decimal) =
-    value > other.value
+    toDouble > other.toDouble
   def <(other: Decimal) =
-    value < other.value
+    toDouble < other.toDouble
   def <=(other: Decimal) =
-    value <= other.value
+    toDouble <= other.toDouble
   def >=(other: Decimal) =
-    value >= other.value
+    toDouble >= other.toDouble
 
-  def +(other: Decimal) =
-    Decimal(value + other.value)
+  def abs: Decimal =
+    Decimal.abs(this)
+  def pow(exponent: Int): Decimal =
+    Decimal.pow(this, exponent)
 
-  def -(other: Decimal) =
-    Decimal(value - other.value)
+  def +(other: Decimal): Decimal =
+    Decimal.add(this, other)
 
-  def *(other: Decimal) =
-    Decimal(value * other.value)
+  def -(other: Decimal): Decimal =
+    Decimal.subtract(this, other)
 
-  def /(other: Decimal) =
-    Decimal(value / other.value)
+  def *(other: Decimal): Decimal =
+    Decimal.multiply(this, other)
+
+  def /(other: Decimal): Decimal =
+    Decimal.divide(this, other)
+}
+
+case class DoubleDecimal(toDouble: Double) extends Decimal
```

Comment:
you know... I don't have great intuition here for how much this would help, but I THINK if you make `Decimal` extend `Any`:

```scala
sealed trait Decimal extends Any with Serializable
```

then this can be a value class:

```scala
case class DoubleDecimal(toDouble: Double) extends AnyVal with Decimal
```

Here are the docs: https://docs.scala-lang.org/overviews/core/value-classes.html

I THINK this will avoid allocation unless you call one of the methods, which we almost certainly will do. I can't understand what the answer to this question is suggesting, but it seems like it's possible to avoid that penalty: https://stackoverflow.com/questions/47140707/value-classes-universal-traits-and-the-necessity-of-instantiation?rq=1